crates/handlers/src/upstream_oauth2/callback.rs

```diff
@@ -152,23 +155,38 @@ pub(crate) async fn handler(
     State(encrypter): State<Encrypter>,
     State(keystore): State<Keystore>,
     State(client): State<reqwest::Client>,
+    State(templates): State<Templates>,
+    PreferredLanguage(locale): PreferredLanguage,
     cookie_jar: CookieJar,
     Path(provider_id): Path<Ulid>,
     query_params: Option<Query<Params>>,
     form_params: Option<Form<Params>>,
-) -> Result<impl IntoResponse, RouteError> {
+) -> Result<Response, RouteError> {
     let provider = repo
         .upstream_oauth_provider()
         .lookup(provider_id)
         .await?
         .filter(UpstreamOAuthProvider::enabled)
         .ok_or(RouteError::ProviderNotFound)?;
 
+    let sessions_cookie = UpstreamSessionsCookie::load(&cookie_jar);
+
     // Read the parameters from the query or the form, depending on what
     // response_mode the provider uses
     let params = match (provider.response_mode, query_params, form_params) {
-        (UpstreamOAuthProviderResponseMode::Query, Some(query_params), None) => query_params.0,
-        (UpstreamOAuthProviderResponseMode::FormPost, None, Some(form_params)) => form_params.0,
+        (UpstreamOAuthProviderResponseMode::Query, Some(Query(query_params)), None) => query_params,
+        (UpstreamOAuthProviderResponseMode::FormPost, None, Some(Form(form_params))) => {
+            // We got there from a cross-site form POST, so we need to render a form with
+            // the same values, which posts back to the same URL
+            if sessions_cookie.is_empty() {
```

Comment:
might be also worth adding a URL parameter or something to indicate the redirect has already been done, so we don't cause a redirect loop if the cookie is genuinely empty for some reason

Reply:
Have done that